File: frontend/src/live-envelope.ts
/**
 * The ``final`` WebSocket envelope — the renderer's HALF of one contract.
 *
 * The backend's half is ``backend/live_envelope.py``, and it is the only
 * place a ``final`` message is built. This module is the only place one
 * is read. Before B-038 there was no "only place" on either side: the
 * backend wrote the message out by hand at five call sites in three
 * mutually incompatible shapes (the Deepgram drain, the local assist
 * with its FLAT coverage keys, and two error paths carrying a third
 * subset), and ``parseLiveWsMessage`` had to be the union of all three.
 * Adding a field to one side was then a silent no-op on the other —
 * which, for a message that IS the delivered transcript, means words
 * quietly not arriving.
 *
 * What keeps the two halves honest is not this comment. It is
 * ``contracts/live-final-envelope.json``: produced by the backend's own
 * builder in ``backend/tests/test_live_envelope.py`` and parsed by
 * ``frontend/tests/live-envelope.test.ts``. Rename or re-nest a field on
 * either side and the other side's suite goes red.
 *
 * Reading rules, unchanged from the parser this replaces and stated
 * once here instead of per field:
 *
 *  * a number is adopted only when it is finite and non-negative;
 *    anything else reads as ABSENT, never as zero. "No field" and
 *    "proven zero" are different facts, and confusing them is how a
 *    false "fully covered" signal gets believed
 *    (BUGS_AUDIT_2026-09-03 §3.4);
 *  * ``coverage`` is adopted only when the backend actually sent the
 *    ``complete`` boolean. A malformed or null block is absent, and an
 *    absent one refuses adoption in ``./live-coverage`` — never the
 *    other way round;
 *  * segments are normalised by the caller's own normaliser, injected,
 *    because the renderer's segment shape is the renderer's business
 *    and this module has no DOM and no globals.
 */

import type { LiveCoverageReport } from "./live-coverage";

/** ``source`` values the backend can put on an envelope. */
export const DEEPGRAM_LIVE_SOURCE = "deepgram-live";
export const LOCAL_ASSIST_SOURCE = "local-assist";

/**
 * The slice of the backend's ``stats`` object the renderer reads.
 *
 * Every field here has a reader in the ``[trace stopLive] FINAL`` line:
 * how the recording was decoded (one Nova-3 stream or two, and what the
 * second one contributed), what the connect and the finalize cost, and
 * what the backend re-decoded from its own spool to make the envelope
 * complete. None of it is visible in the transcript, and together it is
 * the whole diagnosis of a stop in a support log.
 *
 * The wire names are the backend's, verified against the code that
 * writes them (``DeepgramLiveStats.as_dict``, ``backend/deepgram_dual``,
 * ``RecoveryReport.as_dict``). Anything else in ``stats`` is the
 * backend's own bookkeeping and is not parsed.
 */
export interface LiveFinalStats {
  /** ``stats.connect_ms`` — what opening the upstream socket cost. */
  connectMs?: number;
  /** ``stats.finalize_ms`` — what flushing it at stop cost. */
  finalizeMs?: number;
  /** ``stats.dual_stream`` — was this recording read by two streams? */
  dualStream?: boolean;
  /** ``stats.dual_secondary_language`` — the second stream's language. */
  dualSecondaryLanguage?: string;
  /** Words the merge took from the second reading. */
  dualFilledFromSecondary?: number;
  /** Words the merge took from the primary where the second had none. */
  dualFilledFromPrimary?: number;
  /** Seconds of the recording the backend re-decoded to close a gap. */
  recoverySpansSec?: number;
  /** What that re-decode cost, in ms — part of the announced budget. */
  recoveryMs?: number;
  /** Words the re-decode actually spliced into the transcript. */
  recoveryWords?: number;
}

/** The envelope, minus its ``type`` tag. One shape, every field. */
export interface LiveFinalEnvelopeFields<Segment> {
  text: string;
  segments: Segment[];
  durationSec: number;
  source: string;
  error?: string;
  /**
   * The local assist's window report. ``null`` on the wire for every
   * transport that has no notion of windows (a Deepgram stream decodes
   * a continuous socket), absent here for the same and for a malformed
   * one — ``./live-coverage`` refuses to adopt a transcript whose
   * coverage it cannot read, which is the safe direction.
   */
  coverage?: LiveCoverageReport;
  /**
   * Seconds where the backend's own interims recognised speech that no
   * final segment covers. Diagnostic only: the backend closes those
   * spans itself, from its own audio spool, before the envelope is
   * sent, and reports what it re-decoded in ``stats.recovery``.
   */
  uncoveredSpeechSec?: number;
  /** Seconds of audio the backend actually streamed to the provider. */
  streamedSec?: number;
  /** The last second of audio a DECODER (never an interim) reported. */
  coveredEndSec?: number;
  stats?: LiveFinalStats;
}

/**
 * Adopt a wire number only when it is finite and non-negative.
 * Everything else is absent — see the reading rules in the module doc.
 */
export function parseOptionalNonNegativeNumber(value: unknown): number | undefined {
  const n = Number(value);
  return Number.isFinite(n) && n >= 0 ? n : undefined;
}

/**
 * ``stats``, field by field, each standing on its own so a backend that
 * reports one and not another still gets the one it reports through.
 * Returns ``undefined`` when nothing readable was in there at all.
 */
export function parseLiveFinalStats(raw: unknown): LiveFinalStats | undefined {
  if (!raw || typeof raw !== "object") return undefined;
  const source = raw as Record<string, unknown>;
  const stats: LiveFinalStats = {};

  const connectMs = parseOptionalNonNegativeNumber(source.connect_ms);
  if (connectMs !== undefined) stats.connectMs = connectMs;
  const finalizeMs = parseOptionalNonNegativeNumber(source.finalize_ms);
  if (finalizeMs !== undefined) stats.finalizeMs = finalizeMs;

  if (typeof source.dual_stream === "boolean") stats.dualStream = source.dual_stream;
  if (typeof source.dual_secondary_language === "string") {
    const code = source.dual_secondary_language.trim();
    if (code) stats.dualSecondaryLanguage = code;
  }
  const filledSecondary = parseOptionalNonNegativeNumber(source.dual_filled_from_secondary);
  if (filledSecondary !== undefined) stats.dualFilledFromSecondary = filledSecondary;
  const filledPrimary = parseOptionalNonNegativeNumber(source.dual_filled_from_primary);
  if (filledPrimary !== undefined) stats.dualFilledFromPrimary = filledPrimary;

  const recovery = source.recovery;
  if (recovery && typeof recovery === "object") {
    const spans = recovery as Record<string, unknown>;
    const spansSec = parseOptionalNonNegativeNumber(spans.spans_sec);
    if (spansSec !== undefined) stats.recoverySpansSec = spansSec;
    const ms = parseOptionalNonNegativeNumber(spans.ms);
    if (ms !== undefined) stats.recoveryMs = ms;
    const words = parseOptionalNonNegativeNumber(spans.words);
    if (words !== undefined) stats.recoveryWords = words;
  }

  return Object.keys(stats).length ? stats : undefined;
}

/**
 * The ``coverage`` block, adopted only on an explicit ``complete``
 * boolean. Its four numbers default to 0 once that boolean is present —
 * unlike the top-level diagnostics — because the backend sends all five
 * together or none of them, and a partial block would mean a corrupted
 * message rather than an older build.
 */
export function parseLiveCoverage(raw: unknown): LiveCoverageReport | undefined {
  if (!raw || typeof raw !== "object") return undefined;
  const source = raw as Record<string, unknown>;
  if (typeof source.complete !== "boolean") return undefined;
  return {
    complete: source.complete,
    coveredSec: Math.max(0, Number(source.coveredSec) || 0),
    totalSec: Math.max(0, Number(source.totalSec) || 0),
    droppedSec: Math.max(0, Number(source.droppedSec) || 0),
    uncoveredTailSec: Math.max(0, Number(source.uncoveredTailSec) || 0),
  };
}

/**
 * The stats half of the ``[trace stopLive] FINAL`` line.
 *
 * Every field the parser above adopts is printed here, which is the
 * point: a wire field with no reader is a field nobody notices the
 * backend has stopped sending. A fact the backend did not report reads
 * ``n/a`` — never ``0``, which would claim it was measured and found to
 * be nothing.
 */
export function describeLiveFinalStats(stats: LiveFinalStats | undefined): string {
  const ms = (value: number | undefined): string =>
    value === undefined ? "n/a" : `${value.toFixed(0)}ms`;
  const dual = stats?.dualStream
    ? `1/${stats.dualSecondaryLanguage || "?"}` +
      `/+${stats.dualFilledFromSecondary ?? 0}` +
      `/+${stats.dualFilledFromPrimary ?? 0}`
    : "0";
  const recovery =
    stats?.recoverySpansSec === undefined
      ? "n/a"
      : `${stats.recoverySpansSec.toFixed(2)}s/${ms(stats.recoveryMs)}` +
        `/${stats.recoveryWords ?? 0}w`;
  return (
    `connect=${ms(stats?.connectMs)} finalize=${ms(stats?.finalizeMs)} ` +
    `dual=${dual} recovery=${recovery}`
  );
}

/**
 * Parse a ``final`` message body into the envelope the stop delivers.
 *
 * ``normalizeSegment`` is the caller's segment normaliser; segments it
 * rejects are dropped, which is the same rule every other transcript
 * surface in the renderer follows.
 */
export function parseLiveFinalEnvelope<Segment>(
  obj: Record<string, unknown>,
  normalizeSegment: (raw: unknown) => Segment | null,
): LiveFinalEnvelopeFields<Segment> {
  const rawSegments = Array.isArray(obj.segments) ? obj.segments : [];
  const segments = rawSegments
    .map((segment) => normalizeSegment(segment))
    .filter((segment): segment is Segment => segment !== null);
  const error = typeof obj.error === "string" && obj.error ? obj.error : undefined;
  return {
    text: String(obj.text || ""),
    segments,
    durationSec: Math.max(0, Number(obj.durationSec) || 0),
    source: String(obj.source || ""),
    error,
    coverage: parseLiveCoverage(obj.coverage),
    uncoveredSpeechSec: parseOptionalNonNegativeNumber(obj.uncoveredSpeechSec),
    streamedSec: parseOptionalNonNegativeNumber(obj.streamedSec),
    coveredEndSec: parseOptionalNonNegativeNumber(obj.coveredEndSec),
    stats: parseLiveFinalStats(obj.stats),
  };
}

File: backend/live_envelope.py
"""The ``final`` WebSocket envelope — ONE shape, ONE constructor.

Why this module exists (B-038)
------------------------------
``{"type": "final", ...}`` is the single message the renderer delivers a
transcript from: since the "one owner of the text" change the backend
envelope IS the transcript, and the renderer's only remaining decision
is whether an envelope arrived at all. That message used to be written
out by hand at five call sites in ``backend.main`` in THREE mutually
incompatible shapes:

* the Deepgram drain — ``text/segments/durationSec/source/stats/
  uncoveredSpeechSec/streamedSec/coveredEndSec``;
* the local assist — ``text/segments/durationSec/source`` plus FLAT
  ``complete/coveredSec/totalSec/droppedSec/uncoveredTailSec`` and none
  of the four diagnostics above;
* the two error paths (no API key, connect failure) — a third subset,
  one of them with the diagnostics and one without.

The renderer's parser then had to be a union of the three, and every
field added on one side was a silent no-op on the other. This module is
the answer: every ``final`` message is built HERE, so there is exactly
one shape on the wire and every key is always present.

The shape
---------
Every key below is ALWAYS present. Absence is expressed by a value
(``""``, ``0.0``, ``null``), never by a missing key, because a missing
key is indistinguishable from a version mismatch — and a version
mismatch read as a fact is how a "complete" coverage report gets
believed for a session that never reported one.

===================== ========================================
``type``              always ``"final"``
``source``            who read the audio: ``deepgram-live`` or
                      ``local-assist`` (the constants below)
``text``              the delivered transcript, verbatim
``segments``          committed segments, each carrying its own
                      ``words`` and ``source``
``durationSec``       how long the transcript IS (last segment end)
``coveredEndSec``     how far a DECODER reported (never an interim)
``streamedSec``       audio actually handed to the provider
``uncoveredSpeechSec`` speech still not covered after recovery
``error``             ``""`` when the stop succeeded
``stats``             always an object — see ``live_final_stats``
``coverage``          the local assist's window report, or ``null``
                      for a transport that has no notion of windows
===================== ========================================

``coverage`` is ``null`` rather than absent for the same reason: the
renderer's adoption policy (``frontend/src/live-coverage.ts``) refuses
to adopt a transcript whose coverage it cannot read, and "the field is
null" and "this build does not send the field" must not be two
different-looking answers to that question.

The mirror of this module on the other side is
``frontend/src/live-envelope.ts``. Neither is allowed to drift: the
committed fixture ``contracts/live-final-envelope.json`` is PRODUCED by
this module's builder in ``backend/tests/test_live_envelope.py`` and
CONSUMED by ``frontend/tests/live-envelope.test.ts``, so a field added
or renamed on one side fails the other side's suite.
"""

from __future__ import annotations

from typing import Any, Mapping, Optional, Sequence

#: The wire ``type`` this module builds. Named so no caller writes it.
FINAL_MESSAGE_TYPE = "final"

#: ``source`` for a recording read by the Deepgram streaming endpoint —
#: one stream or two, repaired or not. The renderer prints it on the
#: stop trace and ``live-coverage`` compares against ``local-assist``.
DEEPGRAM_LIVE_SOURCE = "deepgram-live"

#: ``source`` for a recording read by the local live assist.
LOCAL_ASSIST_SOURCE = "local-assist"

#: Every key a ``final`` message carries, in wire order. The tests on
#: both sides read this rather than restating the list.
FINAL_ENVELOPE_KEYS = (
    "type",
    "source",
    "text",
    "segments",
    "durationSec",
    "coveredEndSec",
    "streamedSec",
    "uncoveredSpeechSec",
    "error",
    "stats",
    "coverage",
)

#: Every key ``coverage`` carries when it is not ``null``.
COVERAGE_KEYS = (
    "complete",
    "coveredSec",
    "totalSec",
    "droppedSec",
    "uncoveredTailSec",
)


def _as_float(value: Any, default: float = 0.0) -> float:
    """Coerce a wire number, never raising and never producing NaN."""
    try:
        out = float(value)
    except (TypeError, ValueError):
        return default
    if out != out or out in (float("inf"), float("-inf")):  # NaN / ±inf
        return default
    return out


def _non_negative(value: Any) -> float:
    return round(max(0.0, _as_float(value)), 3)


def live_final_stats(
    raw: Optional[Mapping[str, Any]] = None,
    *,
    recovery: Optional[Mapping[str, Any]] = None,
) -> dict:
    """Normalise the ``stats`` object of a ``final`` envelope.

    The session's own ``DeepgramLiveStats.as_dict`` and the dual reader's
    additions pass through unchanged — this backend owns what it reports
    about itself, and the renderer reads only the handful of keys it
    documents (``connect_ms``, ``finalize_ms``, ``dual_stream``,
    ``dual_secondary_language``, ``dual_filled_from_secondary``,
    ``dual_filled_from_primary``, ``recovery``). The one guarantee made
    here is that ``stats`` is an OBJECT on every path, including the two
    error paths that used to omit it entirely, so the renderer never has
    to distinguish "no stats" from "a stop that reported none".
    """
    stats: dict = dict(raw or {})
    if recovery is not None:
        stats["recovery"] = dict(recovery)
    return stats


def live_coverage(
    *,
    complete: bool,
    covered_sec: Any,
    total_sec: Any,
    dropped_sec: Any,
    uncovered_tail_sec: Any,
) -> dict:
    """The local assist's window report, as it travels on the wire.

    Its five numbers used to be FLAT top-level keys on the local-assist
    envelope only — the second of the three shapes B-038 names. They are
    one fact about one transport, so they travel as one object.
    """
    return {
        "complete": bool(complete),
        "coveredSec": _non_negative(covered_sec),
        "totalSec": _non_negative(total_sec),
        "droppedSec": _non_negative(dropped_sec),
        "uncoveredTailSec": _non_negative(uncovered_tail_sec),
    }


def live_final_envelope(
    *,
    source: str,
    text: str = "",
    segments: Optional[Sequence[Mapping[str, Any]]] = None,
    duration_sec: Any = 0.0,
    covered_end_sec: Any = 0.0,
    streamed_sec: Any = 0.0,
    uncovered_speech_sec: Any = 0.0,
    error: str = "",
    stats: Optional[Mapping[str, Any]] = None,
    coverage: Optional[Mapping[str, Any]] = None,
) -> dict:
    """Build the ONE ``final`` message shape. The only constructor.

    Callers pass what they know; everything else takes its documented
    "nothing to report" value rather than vanishing from the payload.
    Segments are passed through as they were built (each carries its own
    ``words`` and ``source``) — placing a word by its time is the
    backend's job and the renderer does not re-derive it.
    """
    return {
        "type": FINAL_MESSAGE_TYPE,
        "source": str(source or ""),
        "text": str(text or ""),
        "segments": [dict(seg) for seg in (segments or [])],
        "durationSec": _non_negative(duration_sec),
        "coveredEndSec": _non_negative(covered_end_sec),
        "streamedSec": _non_negative(streamed_sec),
        "uncoveredSpeechSec": _non_negative(uncovered_speech_sec),
        "error": str(error or ""),
        "stats": live_final_stats(stats),
        "coverage": dict(coverage) if coverage is not None else None,
    }


def envelope_from_result(
    result: Mapping[str, Any],
    *,
    source: str,
    error: str = "",
    coverage: Optional[Mapping[str, Any]] = None,
) -> dict:
    """Build the envelope from a session result dict.

    ``DeepgramLiveSession.drain_transcript`` and ``.partial_result``
    return the SAME dict shape (C6: a partial and a full result may never
    describe the same audio differently), and so does the repaired
    payload ``backend.deepgram_recovery.run_recovery`` produces. One
    adapter for all three keeps the field-by-field copying that used to
    stand at each call site — and used to differ between them — in one
    place.
    """
    return live_final_envelope(
        source=source,
        text=result.get("text", ""),
        segments=result.get("segments") or [],
        duration_sec=result.get("durationSec", 0.0),
        covered_end_sec=result.get("coveredEndSec", 0.0),
        streamed_sec=result.get("streamedSec", 0.0),
        uncovered_speech_sec=result.get("uncoveredSpeechSec", 0.0),
        error=error or str(result.get("error") or ""),
        stats=result.get("stats"),
        coverage=coverage if coverage is not None else result.get("coverage"),
    )


__all__ = [
    "COVERAGE_KEYS",
    "DEEPGRAM_LIVE_SOURCE",
    "FINAL_ENVELOPE_KEYS",
    "FINAL_MESSAGE_TYPE",
    "LOCAL_ASSIST_SOURCE",
    "envelope_from_result",
    "live_coverage",
    "live_final_envelope",
    "live_final_stats",
]
